Import

byorgey opened the pull request:
This still does not work at all, but it's inching closer and I figured it was worth opening a draft PR, so I can start working on a comprehensive PR description in parallel with finishing it up.

xsebek commented:
I tried to test it on the move tutorial:
```diff
   - name: 1P wall
     system: true
     program: |
-      def main = \a. pure noop end
       instant {
-        run "scenarios/Tutorials/move_surveil.sw";
+        import "scenarios/Tutorials/move_surveil.sw" in
         main [entity="wall", room=1]
       }
```
But the robots fail to run:
```
0:00:00.0: Fatal error: Import not found:
/Users/xsebek/Git/swarm-import/scenarios/Tutorials/move_surveil.sw
Please report this as a bug at <https://github.com/swarm-game/swarm/issues/new>.
```

Also, could you please fix the `Swarm/Doc/Pedagogy.hs`? Since it fails to build I cannot run integration tests or `cabal install`. So I run this:
```
cabal run swarm:exe:swarm -- --scenario Tutorials/Move --run <(echo 'move;move') --debug=all
```

byorgey commented:
> But the robots fail to run:

@xsebek what did you do to get that crash?  When I try it, the system robots in the move tutorial do not work properly, but at least it doesn't crash.  But I'm not sure if that's because I'm doing something differently than you, or if I incidentally fixed the crash in the meantime.

xsebek commented:
@byorgey I used import in Move tutorial thus:

<details>
  <summary>Click to expand diff</summary>

```diff
diff --git a/data/scenarios/Tutorials/move.yaml b/data/scenarios/Tutorials/move.yaml
index 5a20eb6e..ea7f35cc 100644
--- a/data/scenarios/Tutorials/move.yaml
+++ b/data/scenarios/Tutorials/move.yaml
@@ -149,54 +149,44 @@ robots:
   - name: 1P wall
     system: true
     program: |
-      def main = \a. pure noop end
       instant {
-        run "scenarios/Tutorials/move_surveil.sw";
+        import "scenarios/Tutorials/move_surveil.sw" in
         main [entity="wall", room=1]
       }
   - name: 2P wall
     system: true
     program: |
-      def main = \a. noop end
       instant {
-        run "scenarios/Tutorials/move_surveil.sw";
+        import "scenarios/Tutorials/move_surveil.sw" in
         main [entity="wall", room=2]
       }
   - name: 3P wall
     system: true
     program: |
-      def main = \a. noop end
-      instant {
-        run "scenarios/Tutorials/move_surveil.sw";
-        main [entity="wall", room=3]
-      }
+      import "scenarios/Tutorials/move_surveil.sw" in
+      main [entity="wall", room=3]
   #################
   ## GARDENERS   ##
   #################
   - name: 1P flower
     system: true
     program: |
-      def main = \a. noop end
       instant {
-        run "scenarios/Tutorials/move_surveil.sw";
+        import "scenarios/Tutorials/move_surveil.sw" in
         main [entity="flower", room=1]
       }
   - name: 2P flower
     system: true
     program: |
-      def main = \a. noop end
       instant {
-        run "scenarios/Tutorials/move_surveil.sw";
+        import "scenarios/Tutorials/move_surveil.sw" in
         main [entity="flower", room=2]
       }
   - name: 3P flower
     system: true
     program: |
-      def main = \a. noop end
-      instant {
-        run "scenarios/Tutorials/move_surveil.sw";
-        main [entity="flower", room=3]
-      }
+      import "scenarios/Tutorials/move_surveil.sw" in
+      main [entity="flower", room=3]
 entities:
   - name: door
     display:
```

</details>

Then run:
```bash
cabal run swarm:exe:swarm -- --scenario Tutorials/Move --run <(echo 'move;move') --debug=all --paused
```
After one step, you should see the robots all turn idle in the robot panel. Select any robot to view their log and you should see the "crash".

byorgey commented:
This is a weird one... the tests pass on Linux and Mac OS; on Windows, testing all the scenario solutions (including `Tutorials/move`) work fine, but we get a failure when testing for "no errors at loading time" saying that the `move` tutorial is missing.  Presumably this has something to do with the fact that the `move` tutorial is the only one that includes an `import`.  It says `import "blah.sw"`, and it is supposed to look for the `.sw` file in the same directory where the `.yaml` file was located, but maybe that fails for some reason...

byorgey commented:
Oops, caching environments does not work, since they can contain references to memory cells in the store, which are not cached (indeed, it doesn't really make sense to cache them).  I guess I should just take out the environment caching for now...

byorgey commented:
I added a few more scenarios that I switched to use `import`, which uncovered a few bugs (now fixed):

- The `farming` scenario used `import` in the `solution`, which wasn't being parsed in a way that it could correctly find relative imports.
- The `roadway` scenario revealed that caching environments would not work, as explained above.

I also created the very beginnings of a standard library in `data/lib/`, at least a standard library we can use in scenarios - not sure it makes sense to advertise or provide it to users.

byorgey commented:
I don't understand why the MacOS build is suddenly failing: https://github.com/swarm-game/swarm/actions/runs/20659588011/job/59319080169  I reran it twice with the same result.

xsebek reviewed (commented):
Great* work @byorgey!

I have gone through the first 100 files out of 168, so about 60% done. 😄 

The import process we have already gone over in the caching PR and talked about it on a call, so no surprises there. 

Most of the modules are just adding the phase parameter. A few robot lenses seem to cause most of the noise in Step/Const, so maybe those would be worth specializing.

I will finish the rest when I get home, but I don't expect I will find much to nitpick. 😄

<sub>*) So great in fact, that it froze the new review UI. I tried writing a comment and it took 10 seconds for each character. Switching to the old one worked, though it collapsed almost every file. 😄 </sub>

xsebek commented on test/unit/TestLanguagePipeline.hs:
```diff
@@ -741,7 +744,7 @@ testLanguagePipeline =
             ( process
                 "def x = 3;\n def y = 3 end;\n def z = 3 end"
-                "3:15:\n  |\n3 |  def z = 3 end\n  |               ^\nunexpected end of input\nexpecting \"!=\", \"&&\", \"()\", \"++\", \"<=\", \"==\", \">=\", \"def\", \"false\", \"let\", \"require\", \"requirements\", \"stock\", \"true\", \"tydef\", \"||\", '\"', '$', '(', '*', '+', '-', '.', '/', ':', ';', '<', '>', '@', '[', '\\', '^', 'end' keyword for definition of 'x', '{', built-in user  ...
+                "3:15:\n  |\n3 |  def z = 3 end\n  |               ^\nunexpected end of input\nexpecting \"!=\", \"&&\", \"()\", \"++\", \"<=\", \"==\", \">=\", \"def\", \"false\", \"import\", \"let\", \"require\", \"requirements\", \"stock\", \"true\", \"tydef\", \"||\", '\"', '$', '(', '*', '+', '-', '.', '/', ':', ';', '<', '>', '@', '[', '\\', '^', 'end' keyword for definition of 'x', '{', bu ...
```
As a side note, that _"'end' keyword for definition of 'x'"_ is so far back I never noticed it before. 😄

xsebek reviewed (approved):
OK, all checked! 👀 

I have a few comments, but nothing that would preclude merging. 🙂 

I look forward to helping with the follow up issues. 👍
